subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVersionControlSystemFactory.java

```diff
@@ -16,17 +16,85 @@
 
 package org.gradle.vcs.internal;
 
+import org.gradle.cache.CacheRepository;
+import org.gradle.cache.FileLockManager;
+import org.gradle.cache.PersistentCache;
 import org.gradle.vcs.VersionControlSpec;
 import org.gradle.vcs.VersionControlSystem;
+import org.gradle.vcs.VersionRef;
 import org.gradle.vcs.git.internal.GitVersionControlSystem;
 
+import java.io.File;
+import java.util.Set;
+
+import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
+
 public class DefaultVersionControlSystemFactory implements VersionControlSystemFactory {
+    private final CacheRepository cacheRepository;
+
+    DefaultVersionControlSystemFactory(CacheRepository cacheRepository) {
+        this.cacheRepository = cacheRepository;
+    }
+
     @Override
     public VersionControlSystem create(VersionControlSpec spec) {
         // TODO: Register these mappings somewhere
+        VersionControlSystem vcs;
         if (spec instanceof DirectoryRepositorySpec) {
-            return new SimpleVersionControlSystem();
+            vcs = new SimpleVersionControlSystem();
+        } else {
+            vcs = new GitVersionControlSystem();
+        }
+        return new ThreadSafeVersionControlSystem(vcs, cacheRepository);
+    }
+
+    private static final class ThreadSafeVersionControlSystem implements VersionControlSystem {
```

Comment:
I don't feel very strongly about this...

I understand the intention here (this makes version control systems thread safe), but it's not just thread safe.  It's also multi-process safe.  Maybe `LockingVersionControlSystem`?